File: app/src/main/java/com/pr0gramm/app/util/AndroidUtility.java
package com.pr0gramm.app.util;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Looper;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.v4.app.TaskStackBuilder;
import android.support.v4.content.res.ResourcesCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.net.ConnectivityManagerCompat;
import android.text.SpannableStringBuilder;
import android.text.style.BulletSpan;
import android.text.style.LeadingMarginSpan;
import android.text.style.URLSpan;
import android.text.util.Linkify;
import android.view.Display;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewParent;
import android.view.inputmethod.InputMethodManager;
import android.widget.TextView;

import com.akodiakson.sdk.simple.Sdk;
import com.crashlytics.android.Crashlytics;
import com.google.common.base.Optional;
import com.google.common.base.Stopwatch;
import com.google.common.base.Strings;
import com.google.common.base.Supplier;
import com.google.common.base.Throwables;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.pr0gramm.app.BuildConfig;
import com.pr0gramm.app.R;
import com.pr0gramm.app.Settings;
import com.pr0gramm.app.services.UriHelper;
import com.pr0gramm.app.ui.PrivateBrowserSpan;
import com.pr0gramm.app.ui.Truss;

import org.apache.commons.lang3.ClassUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.nio.charset.Charset;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

import butterknife.ButterKnife;
import rx.Observable;
import rx.functions.Action1;
import rx.functions.Action2;
import rx.functions.Func1;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static rx.Observable.empty;
import static rx.Observable.just;

/**
 * Place to put everything that belongs nowhere. Thanks Obama.
 */
public class AndroidUtility {
    private static final Logger logger = LoggerFactory.getLogger("AndroidUtility");

    private static final Pattern RE_USERNAME = Pattern.compile("@[A-Za-z0-9]+");
    private static final Pattern RE_GENERIC_LINK = Pattern.compile("(?:https?://)?(?:www\\.)?pr0gramm\\.com(/(?:new|top|user)/[^\\p{javaWhitespace}]*[0-9])");
    private static final Pattern RE_GENERIC_SHORT_LINK = Pattern.compile("/((?:new|top|user)/[^\\p{javaWhitespace}]*[0-9])");

    private static final Cache<String, Boolean> previousExceptions =
            CacheBuilder.<String, Boolean>newBuilder()
                    .expireAfterWrite(1, TimeUnit.MINUTES)
                    .maximumSize(32)
                    .build();

    private AndroidUtility() {
    }

    /**
     * Gets the height of the action bar as definied in the style attribute
     * {@link R.attr#actionBarSize} plus the height of the status bar on android
     * Kitkat and above.
     *
     * @param context A context to resolve the styled attribute value for
     */
    public static int getActionBarContentOffset(Context context) {
        int offset = 0;
        if (Sdk.isAtLeastKitKat()) {
            offset = getStatusBarHeight(context);
        }

        return getActionBarHeight(context) + offset;
    }

    /**
     * Gets the height of the actionbar.
     */
    public static int getActionBarHeight(Context context) {
        TypedArray arr = context.obtainStyledAttributes(new int[]{R.attr.actionBarSize});
        try {
            return arr.getDimensionPixelSize(0, -1);
        } finally {
            arr.recycle();
        }
    }

    /**
     * Gets the height of the statusbar.
     */
    public static int getStatusBarHeight(Context context) {
        int result = 0;
        Resources resources = context.getResources();
        int resourceId = resources.getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = resources.getDimensionPixelSize(resourceId);
        }

        return result;
    }

    public static String urlDecode(String value, Charset charset) {
        try {
            return URLDecoder.decode(value, charset.name());

        } catch (UnsupportedEncodingException e) {
            throw Throwables.propagate(e);
        }
    }

    public static void checkMainThread() {
        if (Looper.getMainLooper().getThread() != Thread.currentThread())
            throw new IllegalStateException("Must be called from the main thread.");
    }

    public static void checkNotMainThread() {
        if (Looper.getMainLooper().getThread() == Thread.currentThread())
            throw new IllegalStateException("Must not be called from the main thread.");
    }

    public static void logToCrashlytics(Throwable error) {
        if (error == null)
            return;

        try {
            // try to rate limit exceptions.
            String key = String.valueOf(error.toString());
            if (Boolean.TRUE.equals(previousExceptions.getIfPresent(key))) {
                return;
            } else {
                previousExceptions.put(key, true);
            }

            // log to crashlytics for fast error reporting.
            Crashlytics.getInstance().core.logException(error);

        } catch (IllegalStateException ignored) {
            // most certainly crashlytics was not activated.
            logger.warn("Looks like crashlytics was not activated. Here is the error:", error);

        } catch (Throwable err) {
            logger.info("Could not send error to crashlytics", err);
        }
    }

    public static Bundle bundle(String key, String value) {
        Bundle bundle = new Bundle();
        bundle.putString(key, value);
        return bundle;
    }

    public static int dp(Context context, int dpValue) {
        float density = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * density);
    }

    public static boolean isOnMobile(Context context) {
        ConnectivityManager cm = (ConnectivityManager) context
                .getSystemService(Context.CONNECTIVITY_SERVICE);

        return ConnectivityManagerCompat.isActiveNetworkMetered(cm);
    }

    /**
     * Gets the color tinted hq-icon
     */
    public static Drawable getTintentDrawable(Context context, @DrawableRes int drawableId, @ColorRes int colorId) {
        Resources resources = context.getResources();
        Drawable icon = DrawableCompat.wrap(ResourcesCompat.getDrawable(resources, drawableId, null));
        DrawableCompat.setTint(icon, ResourcesCompat.getColor(resources, colorId, null));
        return icon;
    }

    public static Logger logger(Object instance) {
        String suffix = Integer.toString(System.identityHashCode(instance), Character.MAX_RADIX);
        return LoggerFactory.getLogger(instance.getClass().getSimpleName() + "-" + suffix);
    }

    /**
     * Returns a CharSequence containing a bulleted and properly indented list.
     *
     * @param leadingMargin In pixels, the space between the left edge of the bullet and the left edge of the text.
     * @param lines         An array of CharSequences. Each CharSequences will be a separate line/bullet-point.
     */
    public static CharSequence makeBulletList(int leadingMargin, List<? extends CharSequence> lines) {
        Truss sb = new Truss();
        for (int idx = 0; idx < lines.size(); idx++) {
            boolean last = idx == lines.size() - 1;
            CharSequence line = lines.get(idx);

            sb.append(line,
                    new BulletSpan(leadingMargin / 3),
                    new LeadingMarginSpan.Standard(leadingMargin));

            sb.append(last ? "" : "\n");
        }

        return sb.build();
    }

    /**
     * Converts the observable into an optional.
     */
    public static <T> Optional<T> toOptional(Observable<T> observable) {
        T element = observable.toBlocking().singleOrDefault(null);
        return Optional.fromNullable(element);
    }

    /**
     * Calls the given action with the value from the optional, if there it is present.
     */
    public static <T> void ifPresent(Optional<T> optional, Action1<T> action) {
        if (optional.isPresent()) {
            action.call(optional.get());
        }
    }

    public static <T, U> void ifPresent(Optional<T> first, Optional<U> second, Action2<T, U> action) {
        if (first.isPresent() && second.isPresent()) {
            action.call(first.get(), second.get());
        }
    }

    /**
     * Measures the runtime of the given runnable and log it to the provided logger.
     */
    public static void time(Logger logger, String name, Runnable runnable) {
        if (BuildConfig.DEBUG) {
            time(logger, name, () -> {
                runnable.run();
                return null;
            });
        } else {
            runnable.run();
        }
    }

    public static <T> T time(Logger logger, String name, Supplier<T> supplier) {
        if (BuildConfig.DEBUG) {
            Stopwatch watch = Stopwatch.createStarted();
            try {
                return supplier.get();
            } finally {
                logger.info("{} took {}", name, watch);
            }
        } else {
            return supplier.get();
        }
    }

    public static void removeView(View view) {
        if (view != null) {
            ViewParent parent = view.getParent();
            if (parent instanceof ViewGroup) {
                ((ViewGroup) parent).removeView(view);
            }
        }
    }

    public static <R> Observable<R> toObservable(Optional<R> optional) {
        if (optional.isPresent()) {
            return just(optional.get());
        } else {
            return empty();
        }
    }

    @TargetApi(Build.VERSION_CODES.JELLY_BEAN)
    public static void setViewBackground(View view, Drawable drawable) {
        if (Sdk.isAtLeastJellyBean()) {
            view.setBackground(drawable);
        } else {
            view.setBackgroundDrawable(drawable);
        }
    }

    public static File toFile(Uri uri) {
        checkArgument("file".equals(uri.getScheme()), "not a file:// uri");
        return new File(uri.getPath());
    }

    private static final Pattern MALICIOUS_COMMENT_CHARS = Pattern.compile("([\\p{Mn}\\p{Mc}\\p{Me}])[\\p{Mn}\\p{Mc}\\p{Me}]+");

    public static void linkify(TextView view, String content) {
        content = MALICIOUS_COMMENT_CHARS.matcher(content).replaceAll("$1");
        SpannableStringBuilder text = SpannableStringBuilder.valueOf(
                RE_GENERIC_LINK.matcher(content).replaceAll("$1"));

        linkify(view, text);
    }

    public static void linkify(TextView view, SpannableStringBuilder text) {
        Uri base = UriHelper.of(view.getContext()).base();
        String scheme = base.getScheme() + "://";

        Linkify.addLinks(text, Linkify.WEB_URLS);

        Linkify.addLinks(text, RE_USERNAME, scheme, null, (match, url) -> {
            String user = match.group().substring(1);
            return base.buildUpon().path("/user").appendEncodedPath(user).toString();
        });

        Linkify.addLinks(text, RE_GENERIC_SHORT_LINK, scheme, null,
                (match, url) -> base.buildUpon().appendEncodedPath(match.group(1)).toString());

        Settings settings = Settings.of(view.getContext());
        if (settings.useIncognitoBrowser()) {
            URLSpan[] spans = text.getSpans(0, text.length(), URLSpan.class);
            for (URLSpan span : spans) {
                String url = span.getURL();
                if (url.contains("://pr0gramm.com/"))
                    continue;

                int start = text.getSpanStart(span);
                int end = text.getSpanEnd(span);
                int flags = text.getSpanFlags(span);
                text.removeSpan(span);

                text.setSpan(new PrivateBrowserSpan(url), start, end, flags);
            }
        }

        view.setText(text);

        view.setMovementMethod(new NonCrashingLinkMovementMethod());
    }

    /**
     * Find one view. Raise a {@link NullPointerException} if the view can not be found.
     *
     * @param view The view to search in
     * @param id   The id of the view to search
     */
    public static <T extends View> T findView(View view, int id) {
        //noinspection unchecked
        return checkNotNull(ButterKnife.findById(view, id));
    }

    public static int buildVersionCode() {
        if (BuildConfig.DEBUG) {
            return 100;
        } else {
            return BuildConfig.VERSION_CODE;
        }
    }

    public static Animator.AnimatorListener endAction(Runnable action) {
        return new AnimatorListenerAdapter() {
            @Override
            public void onAnimationEnd(Animator animation) {
                action.run();
            }
        };
    }

    public static Func1<Boolean, Boolean> isTrue() {
        return val -> val != null && val;
    }

    public static Func1<Boolean, Boolean> isFalse() {
        return val -> val == null || !val;
    }

    public static <T> Func1<T, Boolean> isNotNull() {
        return val -> val != null;
    }

    public static void hideSoftKeyboard(View view) {
        try {
            InputMethodManager imm = (InputMethodManager) view.getContext()
                    .getSystemService(Context.INPUT_METHOD_SERVICE);

            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        } catch (Exception ignored) {
        }
    }

    @ColorInt
    public static int darken(@ColorInt int color, float amount) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[2] *= (1.f - amount);
        return Color.HSVToColor(hsv);
    }

    public static String getMessageWithCauses(Throwable error) {
        String type = ClassUtils.getShortClassName(error.getClass());
        Throwable cause = error.getCause();

        boolean hasCause = cause != null && error != cause;
        boolean hasMessage = !Strings.isNullOrEmpty(error.getMessage()) &&
                (!hasCause || !error.getMessage().contains(cause.getClass().getSimpleName()));

        if (hasMessage) {
            if (hasCause) {
                return String.format("%s(%s), caused by %s",
                        type, error.getMessage(), getMessageWithCauses(cause));
            } else {
                return String.format("%s(%s)", type, error.getMessage());
            }
        } else {
            if (hasCause) {
                return String.format("%s, caused by %s", type, getMessageWithCauses(cause));
            } else {
                return type;
            }
        }
    }

    public static void recreateActivity(Activity activity) {
        final Intent intent = activity.getIntent();
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        TaskStackBuilder.create(activity)
                .addNextIntentWithParentStack(intent)
                .startActivities();
    }

    @TargetApi(Build.VERSION_CODES.KITKAT)
    public static void applyWindowFullscreen(Activity activity, boolean fullscreen) {
        int flags = 0;
        if (fullscreen) {
            flags |= View.SYSTEM_UI_FLAG_HIDE_NAVIGATION;

            if (Sdk.isAtLeastJellyBean()) {
                flags |= View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                        | View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                        | View.SYSTEM_UI_FLAG_FULLSCREEN;
            }

            if (Sdk.isAtLeastKitKat())
                flags |= View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
        }

        View decorView = activity.getWindow().getDecorView();
        decorView.setSystemUiVisibility(flags);
    }

    @SuppressLint("NewApi")
    public static Point screenSize(Activity activity) {
        Point screenSize = new Point();
        Display display = activity.getWindowManager().getDefaultDisplay();

        if (Sdk.isAtLeastJellyBeanMR1()) {
            display.getRealSize(screenSize);
        } else {
            display.getSize(screenSize);
        }

        return screenSize;
    }

    public static boolean screenIsLandscape(Activity activity) {
        Point size = screenSize(activity);
        return size.x > size.y;
    }
}

File: app/src/main/java/com/pr0gramm/app/util/AndroidUtility.kt
package com.pr0gramm.app.util

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.content.Intent
import android.database.sqlite.SQLiteFullException
import android.graphics.Point
import android.graphics.drawable.Drawable
import android.net.ConnectivityManager
import android.os.Build
import android.os.Looper
import android.text.SpannableStringBuilder
import android.text.style.BulletSpan
import android.text.style.LeadingMarginSpan
import android.util.LruCache
import android.view.View
import android.view.WindowInsets
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.app.TaskStackBuilder
import androidx.core.content.getSystemService
import androidx.core.content.res.ResourcesCompat
import androidx.core.content.res.use
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.net.ConnectivityManagerCompat
import androidx.core.text.inSpans
import androidx.core.view.postDelayed
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.pr0gramm.app.BuildConfig
import com.pr0gramm.app.Logger
import com.pr0gramm.app.R
import com.pr0gramm.app.debugConfig
import com.pr0gramm.app.ui.PermissionHelperDelegate
import com.pr0gramm.app.ui.base.AsyncScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import retrofit2.HttpException
import java.io.IOException
import java.io.PrintWriter
import java.io.StringWriter

/**
 * Place to put everything that belongs nowhere. Thanks Obama.
 */
object AndroidUtility {
    private val logger = Logger("AndroidUtility")

    private val EXCEPTION_BLACKLIST =
        listOf("MediaCodec", "dequeueInputBuffer", "dequeueOutputBuffer", "releaseOutputBuffer", "native_")

    private val cache = LruCache<Int, Unit>(6)

    /**
     * Gets the height of the action bar as definied in the style attribute
     * [R.attr.actionBarSize] plus the height of the status bar on android
     * Kitkat and above.

     * @param context A context to resolve the styled attribute value for
     */
    fun getActionBarContentOffset(context: Context): Int {
        return getStatusBarHeight(context) + getActionBarHeight(context)
    }

    /**
     * Gets the height of the actionbar.
     */
    private fun getActionBarHeight(context: Context): Int {
        context.obtainStyledAttributes(intArrayOf(R.attr.actionBarSize)).use { ta ->
            return ta.getDimensionPixelSize(ta.getIndex(0), -1)
        }
    }

    /**
     * Gets the height of the statusbar.
     */
    fun getStatusBarHeight(context: Context): Int {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val window = activityFromContext(context)?.window
            val rootWindowInsets = window?.decorView?.rootWindowInsets
            if (rootWindowInsets != null) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    // this seems to be always correct on android 11 and better.
                    return rootWindowInsets
                        .getInsetsIgnoringVisibility(WindowInsets.Type.statusBars())
                        .top
                }

                // this works for android 6 and above
                return rootWindowInsets.stableInsetTop
            }
        }

        // use the old code as fallback in case we have a really old android
        // or if we don't have a window or decorView

        var result = 0

        val resources = context.resources
        val resourceId = resources.getIdentifier("status_bar_height", "dimen", "android")
        if (resourceId > 0) {
            result = resources.getDimensionPixelSize(resourceId)
        }

        return result
    }

    fun logToCrashlytics(error: Throwable?, force: Boolean = false) {
        val causalChain = error?.causalChain ?: return

        if (causalChain.containsType<CancellationException>()) {
            return
        }

        if (!force) {
            if (causalChain.containsType<PermissionHelperDelegate.PermissionNotGranted>()) {
                return
            }

            if (causalChain.containsType<IOException>() || causalChain.containsType<HttpException>()) {
                logger.warn(error) { "Ignoring network exception" }
                return
            }

            if (causalChain.containsType<SQLiteFullException>()) {
                logger.warn { "Database is full: $error" }
                return
            }
        }

        try {
            val trace = StringWriter().also { w -> error.printStackTrace(PrintWriter(w)) }.toString()
            if (EXCEPTION_BLACKLIST.any { word -> word in trace }) {
                logger.warn("Ignoring exception", error)
                return
            }

            val errorStr = error.toString()
            if ("connect timed out" in errorStr) {
                return
            }

            // try to rate limit exceptions.
            val key = System.identityHashCode(error)
            if (cache.get(key) != null) {
                return
            } else {
                cache.put(key, Unit)
            }

            ignoreAllExceptions {
                FirebaseCrashlytics.getInstance().recordException(error)
            }

        } catch (err: Throwable) {
            logger.warn(err) { "Could not send error $error to crash tool" }
        }
    }

    fun isOnMobile(context: Context?): Boolean {
        context ?: return false

        debugOnly {
            return true
        }

        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        return ConnectivityManagerCompat.isActiveNetworkMetered(cm)
    }

    /**
     * Returns a CharSequence containing a bulleted and properly indented list.

     * @param leadingMargin In pixels, the space between the left edge of the bullet and the left edge of the text.
     * *
     * @param lines         An array of CharSequences. Each CharSequences will be a separate line/bullet-point.
     */
    fun makeBulletList(leadingMargin: Int, lines: List<CharSequence>): CharSequence {
        return SpannableStringBuilder().apply {
            for (idx in lines.indices) {
                inSpans(BulletSpan(leadingMargin / 3)) {
                    inSpans(LeadingMarginSpan.Standard(leadingMargin)) {
                        append(lines[idx])
                    }
                }

                val last = idx == lines.lastIndex
                append(if (last) "" else "\n")
            }
        }
    }

    fun buildVersionCode(): Int {
        return if (BuildConfig.DEBUG) {
            debugConfig.versionOverride ?: BuildConfig.VERSION_CODE
        } else {
            BuildConfig.VERSION_CODE
        }
    }

    fun showSoftKeyboard(view: EditText?) {
        view?.postDelayed(100) {
            try {
                view.requestFocus()

                val imm = view.context.getSystemService<InputMethodManager>()
                imm?.showSoftInput(view, InputMethodManager.SHOW_IMPLICIT)

            } catch (err: Exception) {
                logger.warn(err) { "Failed to show soft keyboard" }
            }
        }
    }

    fun hideSoftKeyboard(view: View?) {
        if (view != null) {
            try {
                val imm = view.context.getSystemService<InputMethodManager>()
                imm?.hideSoftInputFromWindow(view.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)

            } catch (err: Exception) {
                logger.warn(err) { "Failed to hide soft keyboard" }
            }
        }
    }

    fun recreateActivity(activity: Activity) {
        val intent = Intent(activity.intent)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        TaskStackBuilder.create(activity)
            .addNextIntentWithParentStack(intent)
            .startActivities()
    }

    fun applyWindowFullscreen(activity: Activity, fullscreen: Boolean) {
        var flags = 0

        if (fullscreen) {
            flags = flags or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION

            flags = flags or (View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                    or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
                    or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                    or View.SYSTEM_UI_FLAG_FULLSCREEN)

            flags = flags or View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
        }

        val decorView = activity.window.decorView
        decorView.systemUiVisibility = flags
    }

    fun screenSize(activity: Activity): Point {
        val screenSize = Point()
        val display = activity.windowManager.defaultDisplay

        display.getRealSize(screenSize)

        return screenSize
    }

    fun screenIsLandscape(activity: Activity?): Boolean {
        if (activity == null) {
            return false
        }

        val size = screenSize(activity)
        return size.x > size.y
    }

    /**
     * Tries to get a basic activity from the given context. Returns an empty observable,
     * if no activity could be found.
     */
    fun activityFromContext(context: Context): Activity? {
        if (context is Activity)
            return context

        if (context is ContextWrapper)
            return activityFromContext(context.baseContext)

        return null
    }

    @ColorInt
    fun resolveColorAttribute(context: Context, attr: Int): Int {
        val arr = context.obtainStyledAttributes(intArrayOf(attr))
        try {
            return arr.getColor(arr.getIndex(0), 0)
        } finally {
            arr.recycle()
        }
    }
}

@Suppress("NOTHING_TO_INLINE")
inline fun checkMainThread() = debugOnly {
    if (Looper.getMainLooper().thread !== Thread.currentThread()) {
        Logger("AndroidUtility").error { "Expected to be in main thread but was: ${Thread.currentThread().name}" }
        throw IllegalStateException("Must be called from the main thread.")
    }
}

@Suppress("NOTHING_TO_INLINE")
inline fun checkNotMainThread(msg: String? = null) = debugOnly {
    if (Looper.getMainLooper().thread === Thread.currentThread()) {
        Logger("AndroidUtility").error { "Expected not to be on main thread: $msg" }
        throw IllegalStateException("Must not be called from the main thread: $msg")
    }
}

inline fun <T> doInBackground(crossinline action: suspend () -> T): Job {
    return AsyncScope.launch {
        try {
            action()
        } catch (thr: Throwable) {
            // log it
            AndroidUtility.logToCrashlytics(BackgroundThreadException(thr))
        }
    }
}

class BackgroundThreadException(cause: Throwable) : RuntimeException(cause)

fun Throwable.getMessageWithCauses(): String {
    val error = this
    val type = javaClass.name
        .replaceFirst(".+\\.".toRegex(), "")
        .replace('$', '.')

    val cause = error.cause

    val hasCause = cause != null && error !== cause
    val message = error.message ?: ""
    val hasMessage = message.isNotBlank() && (
            !hasCause || (cause != null && cause.javaClass.directName !in message))

    return if (hasMessage) {
        if (hasCause && cause != null) {
            "$type(${error.message}), caused by ${cause.getMessageWithCauses()}"
        } else {
            "$type(${error.message})"
        }
    } else {
        if (hasCause && cause != null) {
            "$type, caused by ${cause.getMessageWithCauses()}"
        } else {
            type
        }
    }
}

/**
 * Gets the color tinted hq-icon
 */
fun Context.getTintedDrawable(@DrawableRes drawableId: Int, @ColorRes colorId: Int): Drawable {
    val icon = DrawableCompat.wrap(AppCompatResources.getDrawable(this, drawableId)!!.mutate())
    DrawableCompat.setTint(icon, ResourcesCompat.getColor(resources, colorId, null))
    return icon
}

fun ImageView.setImageResource(@DrawableRes drawableId: Int, @ColorRes colorId: Int) {
    setImageDrawable(context.getTintedDrawable(drawableId, colorId))
}
